File: daemon-mac/Sources/FocusLockDaemon/FamilyMessages.swift
import Foundation

// ── On-disk + IPC family state ─────────────────────────────────────────────

/// Persisted under /Library/Application Support/FocusLock/family.json.
/// Missing/empty file means not paired.
struct FamilyConfig: Codable {
    var serverUrl: String
    var accountId: String
    var deviceId: String
    var deviceToken: String  // 1-year JWT
    var pairedAt: String
    var hostname: String?
    /// Mirror of the Windows flag. macOS daemon doesn't currently implement
    /// firewall lockdown (pfctl deferred), but the flag round-trips so a
    /// future implementation can read it without a schema bump.
    var firewallLockdownEnabled: Bool?
}

struct FamilyRuleSummary: Codable {
    var id: String
    var kind: String          // block_now | schedule | unblock_all | unblock_specific
    var targetApps: [String]
    var targetDomains: [String]
    var scheduleCron: String?
    var createdAt: String
    var expiresAt: String?
}

struct FamilyStatus: Codable {
    var paired: Bool
    var connected: Bool
    var accountId: String?
    var deviceId: String?
    var serverUrl: String?
    var lastConnectedAt: String?
    var lastDisconnectedAt: String?
    var lastError: String?
    var activeRuleCount: Int
    /// Seconds since the last successful connection. 0 when currently connected.
    var offlineSeconds: Int
    var activeRules: [FamilyRuleSummary]
    var firewallLockdownEnabled: Bool
    var firewallLockdownActive: Bool
}

// ── IPC payloads ───────────────────────────────────────────────────────────

struct FamilyRedeemPayload: Codable {
    var code: String
    var serverUrl: String
}

struct FamilyRedeemResult: Codable {
    var accountId: String
    var deviceId: String
    var pairedAt: String
}

/// Snapshot of the host environment as seen by the daemon. Used by the parent
/// setup flow to gate pairing on a non-admin child account being available
/// and to surface elevation warnings.
struct FamilyEnvironment: Codable {
    var platform: String            // "windows" | "macos"
    var osVersion: String
    var daemonElevated: Bool        // SYSTEM (Win) or root (mac)
    var uacEnabled: Bool?           // Windows-only; nil on macOS
    var currentUser: String?        // Daemon's own identity (informational)
    var localUsers: [LocalUserAccount]  // Per-user admin enumeration; empty when unavailable
}

/// One local account on the host. Surfaced so the parent setup flow can name
/// which accounts need to be demoted from administrator before pairing.
struct LocalUserAccount: Codable {
    var name: String
    var isAdmin: Bool
    var isCurrent: Bool
    var isBuiltIn: Bool
}

// ── Cloud wire shapes ──────────────────────────────────────────────────────

struct CloudRule: Codable {
    var id: String
    var deviceId: String
    var kind: String
    var targetApps: [String]
    var targetDomains: [String]
    var scheduleCron: String?
    var active: Bool
    var createdAt: String
    var expiresAt: String?
}

struct CloudRulesEnvelope: Codable {
    var rules: [CloudRule]
}

struct CloudMessage: Codable {
    var type: String
    var rule: CloudRule?
    var ruleId: String?
    var t: Int64?
}

struct RedeemResponse: Codable {
    var accountId: String
    var deviceId: String
    var deviceToken: String
    var expiresInSeconds: Int64
}

// ── Family approval requests (Phase 3.2) ─────────────────────────────────────

struct RequestUnblockPayload: Codable {
    var target: String
    var targetKind: String      // "app" | "domain"
    var minutes: Int            // 15 | 30 | 60
}

/// v1.4.1: either a freshly-created request (success), or a typed anti-spam
/// conflict from the family-server. `conflictCode` is nil on success.
/// Exactly one of `pendingRequestId` / `retryAfter` is set per code.
struct RequestUnblockResult: Codable {
    var requestId: String      // empty on conflict
    var expiresAt: String      // empty on conflict

    var conflictCode:     String?   // "pending_exists" | "deny_cooldown" | nil
    var pendingRequestId: String?
    var retryAfter:       String?

    init(requestId: String, expiresAt: String,
         conflictCode: String? = nil,
         pendingRequestId: String? = nil,
         retryAfter: String? = nil) {
        self.requestId        = requestId
        self.expiresAt        = expiresAt
        self.conflictCode     = conflictCode
        self.pendingRequestId = pendingRequestId
        self.retryAfter       = retryAfter
    }
}

struct RequestStatusPayload: Codable {
    var requestId: String
}

struct RequestStatusResult: Codable {
    var status: String          // "pending" | "approved" | "denied" | "expired"
    var resolutionRuleExpiresAt: String?
}

File: daemon-win/FocusLock.Daemon/Models/FamilyMessages.cs
using System.Text.Json.Serialization;

namespace FocusLock.Daemon.Models;

// ── On-disk + IPC family state ─────────────────────────────────────────────

/// <summary>
/// Persisted under ProgramData\FocusLock\family.json. Holds the parent-account
/// binding for this child device. Empty/missing file means not paired.
/// </summary>
public sealed class FamilyConfig
{
    public string ServerUrl   { get; set; } = string.Empty;   // e.g. https://family.focus-lock.app
    public string AccountId   { get; set; } = string.Empty;
    public string DeviceId    { get; set; } = string.Empty;
    public string DeviceToken { get; set; } = string.Empty;   // long-lived (1y) JWT
    public string PairedAt    { get; set; } = string.Empty;   // ISO 8601
    public string? Hostname   { get; set; }
    /// <summary>
    /// Opt-in Windows-only "kid pulled the network cable" mitigation. When set,
    /// the daemon applies per-app Windows Firewall outbound-block rules for
    /// cached block_now process targets after the WebSocket has been offline
    /// for > 5 minutes, on top of the existing process-kill loop. Rules
    /// auto-clear on reconnect. See FirewallLockdownService.
    /// </summary>
    public bool FirewallLockdownEnabled { get; set; }
}

public sealed class FamilyStatus
{
    public bool   Paired               { get; set; }
    public bool   Connected            { get; set; }
    public string? AccountId           { get; set; }
    public string? DeviceId            { get; set; }
    public string? ServerUrl           { get; set; }
    public string? LastConnectedAt     { get; set; }
    public string? LastDisconnectedAt  { get; set; }
    public string? LastError           { get; set; }
    public int    ActiveRuleCount     { get; set; }
    /// <summary>Seconds since the last successful connection. 0 when currently connected.</summary>
    public int    OfflineSeconds      { get; set; }
    public List<FamilyRuleSummary> ActiveRules { get; set; } = new();
    public bool   FirewallLockdownEnabled { get; set; }
    public bool   FirewallLockdownActive  { get; set; }
}

public sealed class FamilyRuleSummary
{
    public string Id            { get; set; } = string.Empty;
    public string Kind          { get; set; } = string.Empty; // block_now | schedule | unblock_all | unblock_specific
    public List<string> TargetApps    { get; set; } = new();
    public List<string> TargetDomains { get; set; } = new();
    public string? ScheduleCron       { get; set; }
    public string CreatedAt           { get; set; } = string.Empty;
    public string? ExpiresAt          { get; set; }
}

// ── IPC payloads (UI → daemon) ─────────────────────────────────────────────

public sealed class FamilyRedeemPayload
{
    public string Code      { get; set; } = string.Empty;
    public string ServerUrl { get; set; } = string.Empty;
}

public sealed class FamilyRedeemResult
{
    public string AccountId { get; set; } = string.Empty;
    public string DeviceId  { get; set; } = string.Empty;
    public string PairedAt  { get; set; } = string.Empty;
}

/// <summary>
/// Snapshot of the host environment as seen by the daemon. Used by the parent
/// setup flow to gate pairing on a non-admin child account being available
/// and to surface UAC/elevation warnings.
/// </summary>
public sealed class FamilyEnvironment
{
    public string Platform        { get; set; } = string.Empty; // "windows" | "macos"
    public string OsVersion       { get; set; } = string.Empty;
    public bool   DaemonElevated  { get; set; }                 // SYSTEM (Win) or root (mac)
    public bool?  UacEnabled      { get; set; }                 // Windows only; null elsewhere
    public string? CurrentUser    { get; set; }                 // Daemon's own identity (informational)
    public List<LocalUserAccount> LocalUsers { get; set; } = new(); // Per-user admin enumeration; empty if unavailable
}

/// <summary>
/// One local account on the host. Surfaced so the parent setup flow can name
/// which accounts need to be demoted from administrator before pairing.
/// </summary>
public sealed class LocalUserAccount
{
    public string Name       { get; set; } = string.Empty;
    public bool   IsAdmin    { get; set; }
    /// <summary>True if this account is currently logged in (best-effort).</summary>
    public bool   IsCurrent  { get; set; }
    /// <summary>True for accounts the OS marks as built-in (Administrator, Guest, DefaultAccount, WDAGUtilityAccount, etc.).</summary>
    public bool   IsBuiltIn  { get; set; }
}

// ── Cloud wire shapes (snake_case in transit handled by JSON opts) ─────────

/// <summary>
/// Server-side LockRule shape (camelCase JSON over the wire).
/// </summary>
public sealed class CloudRule
{
    public string Id   { get; set; } = string.Empty;
    public string DeviceId { get; set; } = string.Empty;
    public string Kind { get; set; } = string.Empty;
    public List<string> TargetApps    { get; set; } = new();
    public List<string> TargetDomains { get; set; } = new();
    public string? ScheduleCron       { get; set; }
    public bool   Active              { get; set; } = true;
    public string CreatedAt           { get; set; } = string.Empty;
    public string? ExpiresAt          { get; set; }
}

/// <summary>
/// Discriminated WS message envelope from server. Only the fields we read are
/// modelled; unknown fields are ignored.
/// </summary>
public sealed class CloudMessage
{
    [JsonPropertyName("type")]   public string Type   { get; set; } = string.Empty;
    [JsonPropertyName("rule")]   public CloudRule? Rule { get; set; }
    [JsonPropertyName("ruleId")] public string? RuleId { get; set; }
    [JsonPropertyName("t")]      public long?   T      { get; set; }
}

// ── Family approval requests (Phase 3.2) ─────────────────────────────────────

public sealed class RequestUnblockPayload
{
    public string Target     { get; set; } = string.Empty;
    public string TargetKind { get; set; } = string.Empty;   // "app" | "domain"
    public int    Minutes    { get; set; }                   // 15 | 30 | 60
}

public sealed class RequestUnblockResult
{
    // Success case: requestId + expiresAt are populated, ConflictCode is null.
    public string  RequestId { get; set; } = string.Empty;
    public string  ExpiresAt { get; set; } = string.Empty;

    // v1.4.1 — Anti-spam conflict case: ConflictCode is "pending_exists" or
    // "deny_cooldown". Exactly one of PendingRequestId / RetryAfter is set
    // per code. UI branches on ConflictCode.
    public string? ConflictCode      { get; set; }
    public string? PendingRequestId  { get; set; }
    public string? RetryAfter        { get; set; }
}

public sealed class RequestStatusPayload
{
    public string RequestId { get; set; } = string.Empty;
}

public sealed class RequestStatusResult
{
    public string Status { get; set; } = string.Empty;       // pending | approved | denied | expired
    public string? ResolutionRuleExpiresAt { get; set; }
}
